src/dagParser.cpp

```diff
@@ -0,0 +1,152 @@
+
+    // Tasks
+    nodeInstance.task = nodeJson.at("task").get<std::string>();
+    nodeInstance.name = nodeJson.at("name").get<std::string>();
+    nodeInstance.description = nodeJson.at("description").get<std::string>();
+    nodeInstance.executor = nodeJson.at("executor").get<std::string>();
+
+    // Container
+    nodeInstance.container.image =
+        nodeJson.at("container").at("image").get<std::string>();
+    nodeInstance.container.resources.cpu =
+        nodeJson.at("container").at("resources").at("cpu").get<uint32_t>();
+    nodeInstance.container.resources.memory =
+        nodeJson.at("container").at("resources").at("memory").get<uint32_t>();
+
+    // Scaling parameters
+    nodeInstance.scalingParameters.taskComplexity =
+        nodeJson.at("scalingParameters").at("taskComplexity").get<uint32_t>();
+
+    // Communication task with optional payload size
+    if (nodeJson.contains("communicationTask") &&
+        nodeJson["communicationTask"].contains("payloadSize")) {
+      nodeInstance.communicationTask.payloadSize =
+          nodeJson["communicationTask"]["payloadSize"].get<uint32_t>();
+    }
+
+    // Optional downstream nodes
+    if (nodeJson.contains("downstream")) {
+      std::vector<Edge> edges;
+      for (const auto &edgeJson : nodeJson.at("downstream")) {
+        Edge edgeInstance;
+        edgeInstance.id = edgeJson.at("id").get<std::string>();
+        edgeInstance.target = edgeJson.at("target").get<std::string>();
+        edgeInstance.max_partitions =
+            edgeJson.at("max_partitions").get<uint32_t>();
+        edges.push_back(edgeInstance);
+      }
+      nodeInstance.downstream = edges;
+    }
+
+    nodeList.push_back(nodeInstance);
+  }
+
+  return DAGParser(appName, nodeList);
+}
+
+void DAGParser::printNodeDetails() {
+  std::cout << "Application Name: " << applicationName << std::endl;
+
+  for (const auto &node : nodes) {
+    std::cout << "Task: " << node.task << ", Name: " << node.name << std::endl;
+    std::cout << "  Description: " << node.description << std::endl;
+    std::cout << "  Executor: " << node.executor << std::endl;
+    std::cout << "  Container Image: " << node.container.image << std::endl;
+    std::cout << "  Resources - CPU: " << node.container.resources.cpu
+              << ", Memory: " << node.container.resources.memory << std::endl;
+    std::cout << "  Task Complexity: " << node.scalingParameters.taskComplexity
+              << std::endl;
+
+    if (node.communicationTask.payloadSize.has_value()) {
+      std::cout << "  Payload Size: "
+                << node.communicationTask.payloadSize.value() << std::endl;
+    }
+
+    if (node.downstream.has_value()) {
+      std::cout << "  Downstream: " << std::endl;
+      for (const auto &edge : node.downstream.value()) {
+        std::cout << "    Edge ID: " << edge.id << ", Target: " << edge.target
+                  << ", Max Partitions: " << edge.max_partitions << std::endl;
+      }
+    }
+  }
+}
```

Comment:
I wonder if we should just return a string with the node information from this method instead, or let the method accept a generalized `ostream` parameter where the node details are written to.

Reply:
we can remove this method actually. I was just printing to see the details. It does not have any further usage IMO

Reply:
Okay, that is also fine :)